MaterializedMySQL: Introduce MySQL Binlog Client

valbok opened the pull request:
The MySQL Binlog Client provides a mechanism in ClickHouse to share the binlog from a MySQL instance among multiple MaterializedMySQL databases. This avoids consuming unnecessary bandwidth and CPU when replicating more than one schema/database.

Suggesting to disable this feature by default for now. It should be explicitly enabled by `SETTINGS use_binlog_client=1`.
But if you would permanently enable it in `MaterializedMySQLSettings`, it should keep old behavior and all tests should pass too.

Details:
1. Introduced `IBinlog` and its impl to read the binlog events from socket - `BinlogFromSocket`, or file - `BinlogFromFile`. Based on prev impl of `EventBase` and the same old binlog parsers. It fully keeps BC with old version. Fixed `./check-mysql-binlog` to test new impl.
2. Introduced `BinlogEventsDispatcher`, it reads the event from the source `IBinlog` and sends it to currently attached `IBinlog` instances.
3. Introduced `BinlogClient`, which is used to group a list of `BinlogEventsDispatcher` by MySQL binlog connection which is defined by `user:password@host:port`. All dispatchers with the same binlog position should be merged to one.
4. Introduced `BinlogClientFactory`, which is a singleton and it is used to track all binlogs created over the instance.
5. Introduced `use_binlog_client` setting to `MaterializedMySQL`, which forces to reuse a `BinlogClient` if it already exists in `BinlogClientFactory` or create new one. By default, it is disabled.
6. Introduced `max_bytes_in_binlog_queue` setting to define the limit of bytes in binlog's queue of events. If bytes in the queue increases this limit, `BinlogEventsDispatcher` will stop reading new events from source `IBinlog` until the space for new events will be freed.
7. Introduced `max_milliseconds_to_wait_in_binlog_queue` setting to define max ms to wait when the max bytes exceeded.
8. Introduced `max_bytes_in_binlog_dispatcher_buffer` setting to define max bytes in the binlog dispatcher's buffer before it is flushed to attached binlogs.
9. Introduced `max_flush_milliseconds_in_binlog_dispatcher` setting to define max milliseconds in the binlog dispatcher's buffer to wait before it is flushed to attached binlogs.
10. Introduced `system.mysql_binlogs` system table, which shows a list of active binlogs.
11. Introduced `UnparsedRowsEvent` and `MYSQL_UNPARSED_ROWS_EVENT`, which defines that an event is not parsed and should be explicitly parsed later.

There are some additional improvements:
-  Postponing the parsing of the binlog event. For old implementation the parsing of the event is done every time it is received by a database. Now if the database is using The MySQL Binlog Client, the event is parsed only once and in its the sync thread.
- Since there are many databases for one connection, it needs to deliver all DDL to databases regardless of table filter. But each database can filter them out on the sync thread if it is needed. Not possible to filter tables on Dispatcher's side.
- Parsing schema/table from DDL which is used later to filter it out.
- Fixed UInt32 overflow when Pos > End_log_pos.
- Introduced timestamp to Position, to compare events or seconds behind MySQL if needed. Also added comparison functions.
- Introduced clearing the pending GTID on reconnect. Author: kangxiang
- Fixed a bug when it is not possible to apply DDL since syntax error or unsupported SQL.

@larspars is the author of following:
- `GTIDSets::contains()` - allows to check if the GTID Sets contains another set.
- `ReplicationHelper` - helper to be used in integration tests.
- `shouldReconnectOnException()` - Improved the logic to retry on failure. Allows to reconnect if needed f.e.

    
<!---
A technical comment, you are free to remove or leave it as it is when PR is created
The following categories are used in the next scripts, update them accordingly
utils/changelog/changelog.py
tests/ci/cancel_and_rerun_workflow_lambda/app.py
-->
### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
- Introduced MySQL Binlog Client for MaterializedMySQL: One binlog connection for many databases.

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

<!---
Directly edit documentation source files in the "docs" folder with the same pull-request as code changes

or

Add a user-readable short description of the changes that should be added to docs.clickhouse.com below.

At a minimum, the following information should be added (but add more as needed).
- Motivation: Why is this function, table engine, etc. useful to ClickHouse users?

- Parameters: If the feature being added takes arguments, options or is influenced by settings, please list them below with a brief explanation.

- Example use: A query or command.
-->


> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

valbok commented:
Should it be an Improvement instead of Feature?

robot-clickhouse-ci-2 commented:
<!-- automatic status comment for PR #57323 from valbok/ClickHouse:mysql-binlog-client -->
*This is an automated comment for commit 362ada592fbe02fdcced127a0fd472437fd3069c with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/57323/362ada592fbe02fdcced127a0fd472437fd3069c/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td><a href="https://github.com/ClickHouse/ClickHouse/actions/runs/7030020889">✅ success</a></td></tr>
<tbody>
</table>
</details>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/57323/362ada592fbe02fdcced127a0fd472437fd3069c/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

valbok commented:
```
unknown file: Failure
C++ exception with description "Cannot open file /src/Databases/MySQL/tests/data/binlog.001390, errno: 2, strerror: No such file or directory" thrown in the test body.
```

Trying to understand why it could not find the files, could it be because they are not copied/bound?

valbok commented:
https://s3.amazonaws.com/clickhouse-test-reports/57323/24716de7c5041f912d775319edc6a7f46dbe366f/fuzzer_astfuzzerasan/report.html
```2023.12.05 11:41:43.754538 [ 1332 ] {} <Fatal> BaseDaemon: 8. DB::Exception::Exception<unsigned long&, unsigned long const&>(int, FormatStringHelperImpl<std::type_identity<unsigned long&>::type, std::type_identity<unsigned long const&>::type>, unsigned long&, unsigned long const&) @ 0x000000000d6ad95d in /workspace/clickhouse
2023.12.05 11:41:43.898675 [ 1332 ] {} <Fatal> BaseDaemon: 9. ./build_docker/./src/Storages/MergeTree/MergeTreeRangeReader.cpp:1246: DB::MergeTreeRangeReader::continueReadingChain(DB::MergeTreeRangeReader::ReadResult const&, unsigned long&) @ 0x00000000321ced64 in /workspace/clickhouse
2023.12.05 11:41:43.997817 [ 1332 ] {} <Fatal> BaseDaemon: 10. ./build_docker/./src/Storages/MergeTree/MergeTreeRangeReader.cpp:991: DB::MergeTreeRangeReader::read(unsigned long, DB::MarkRanges&) @ 0x00000000321ca4ad in /workspace/clickhouse```

valbok committed:
MaterializedMySQL: Introduce MySQL Binlog Client

One binlog connection for many databases.

Suggesting to disable this feature by default for now. It should be explicitly enabled by SETTINGS use_binlog_client=1.
But if you would permanently enable it in MaterializedMySQLSettings, it should keep old behavior and all tests should pass too.

1. Introduced `IBinlog` and its impl to read the binlog events from socket - `BinlogFromSocket`, or file - `BinlogFromFile`. Based on prev impl of `EventBase` and the same old binlog parsers. It fully keeps BC with old version. Fixed `./check-mysql-binlog` to test new impl.
2. Introduced `BinlogEventsDispatcher`, it reads the event from the source `IBinlog` and sends it to currently attached `IBinlog` instances.
3. Introduced `BinlogClient`, which is used to group a list of `BinlogEventsDispatcher` by MySQL binlog connection which is defined by `user:password@host:port`. All dispatchers with the same binlog position should be merged to one.
4. Introduced `BinlogClientFactory`, which is a singleton and it is used to track all binlogs created over the instance.
5. Introduced `use_binlog_client` setting to `MaterializedMySQL`, which forces to reuse a `BinlogClient` if it already exists in `BinlogClientCatalog` or create new one. By default, it is disabled.
6. Introduced `max_bytes_in_binlog_queue` setting to define the limit of bytes in binlog's queue of events. If bytes in the queue increases this limit, `BinlogEventsDispatcher` will stop reading new events from source `IBinlog` until the space for new events will be freed.
7. Introduced `max_milliseconds_to_wait_in_binlog_queue` setting to define max ms to wait when the max bytes exceeded.
7. Introduced `max_milliseconds_to_wait_in_binlog_queue` setting to define max ms to wait when the max bytes exceeded.
8. Introduced `max_bytes_in_binlog_dispatcher_buffer` setting to define max bytes in the binlog dispatcher's buffer before it is flushed to attached binlogs.
9. Introduced `max_flush_milliseconds_in_binlog_dispatcher` setting to define max milliseconds in the binlog dispatcher's buffer to wait before it is flushed to attached binlogs.
10. Introduced `system.mysql_binlogs` system table, which shows a list of active binlogs.
11. Introduced `UnparsedRowsEvent` and `MYSQL_UNPARSED_ROWS_EVENT`, which defines that an event is not parsed and should be explicitly parsed later.
12. Fixed bug when not possible to apply DDL since syntax error or unsupported SQL.

@larspars is the author of following:
`GTIDSets::contains()`
`ReplicationHelper`
`shouldReconnectOnException()`

valbok commented:
Checking why 
```MaterializedMySQLSyncThread: Code: 62. DB::Exception: Syntax error: failed at position 136 ('FLUSH'): FLUSH TABLES.```

valbok commented:
Might be an bug when ClickHouse is blocked and can't be stopped. Checked all materialized sync threads and materialized mysql databases, and they have been exited normally and did not block.
```
WARNING : Stop ClickHouse raised an error Command ['docker', 'exec', '-u', 'root', 'roottestmaterializedmysqldatabase_node1_1', 'bash', '-c', "gdb -batch -ex 'thread apply all bt full' -p 3084 > /ClickHouse/tests/integration/test_materialized_mysql_database/_instances_0/node1/logs/stdout.log"] return non-zero code 1: bash: line 1: /ClickHouse/tests/integration/test_materialized_mysql_database/_instances_0/node1/logs/stdout.log: No such file or directory```

valbok commented:
All test_materialized_mysql_database tests passed
@kssenii

kssenii commented on src/Databases/MySQL/MySQLBinlogClientCatalog.h:
```diff
@@ -0,0 +1,58 @@
+namespace DB::MySQLReplication
+{
+
+class BinlogClientCatalog final : boost::noncopyable
```
```suggestion
class BinlogClientFactory final : boost::noncopyable
```
Seems ok? Too many different names:)

valbok replied:
It does not only create/instantiate the binlog client, it also keeps track of them and returns if it was already existed.

kssenii commented on src/Databases/MySQL/MySQLBinlogEventsDispatcher.cpp:
```diff
@@ -0,0 +1,612 @@
+        std::lock_guard lock(mutex);
+        is_cancelled = true;
+        auto exc = std::make_exception_ptr(Exception(ErrorCodes::UNKNOWN_EXCEPTION, "Dispatcher {} has been already destroyed", logger_name));
+        cleanupLocked([&](const auto & binlog)
```
Seems to be a problem if `cleanupLocked` throws exception, add try - catch?

valbok commented on src/Databases/MySQL/MySQLBinlog.cpp:
```diff
@@ -0,0 +1,512 @@
+    if (payload.eof())
+        throw Exception(ErrorCodes::ATTEMPT_TO_READ_AFTER_EOF, "Attempt to read after EOF.");
+
+    UInt16 header = static_cast<unsigned char>(*payload.position());
```
Sorry, position() should not be changed there, so need to read directly.

valbok commented on src/Core/MySQL/MySQLGtid.cpp:
```diff
@@ -188,4 +188,46 @@ String GTIDSets::toPayload() const
+
+    auto mine = intervals.begin(), other = gtid_set.intervals.begin();
+    auto my_end = intervals.end(), other_end = gtid_set.intervals.end();
+    while (mine != my_end && other != other_end)
```
Unfortunately it is only diff between 2 sets, it is about one set contains another. std::set_difference returns diff for both.

valbok commented on src/Databases/MySQL/MySQLBinlog.cpp:
```diff
@@ -0,0 +1,512 @@
+{
+    if (older.gtid_sets.contains(newer.gtid_sets))
+        return false;
+    for (const auto & new_set : newer.gtid_sets.sets)
```
Unfortunately it needs to confirm that all sets in newer position contains the same UUID from old one. std::set_difference will show difference from both positions. But rewrote it using std::set

kssenii reviewed (approved):
Sorry for the delay - I was on vacation.

kssenii commented:
> Stateless tests (aarch64) — Tests are not finished

https://github.com/ClickHouse/ClickHouse/issues/58508

valbok commented:
Thanks

alexey-milovidov commented:
@valbok, Hi! A test has failed: https://s3.amazonaws.com/clickhouse-test-reports/59166/383ae86ebb0da8a962521f55b08186331eb0f676/integration_tests__asan__analyzer__[4_6].html

Let's fix this test or revert this PR.

alexey-milovidov commented on src/Core/MySQL/tests/gtest_MySQLGtid.cpp:
```diff
@@ -0,0 +1,40 @@
+             contained1, contained2, contained3, contained4, contained5,
+             not_contained1, not_contained2, not_contained3, not_contained4, not_contained5, not_contained6;
+
+    gtid_set.parse("2174B383-5441-11E8-B90A-C80AA9429562:1-3:11:47-49, 24DA167-0C0C-11E8-8442-00059A3C7B00:1-19:47-49:60");
```
This unit test contains incorrect UUIDs.

valbok replied:
Should we try to find correct ones instead?

alexey-milovidov replied:
I already changed it, see the master branch.
